### DISCUSSION TITLE
Can I send Native token(ETH, BNB) and others(USDT, BUSD, DOGE) with one Hook? or Can we send ETH or BNB via useContractWrite()?

### DISCUSSION THREAD

**Discussion by sniper365** [wevm/wagmi#1873] [Q&A]
I tried like below....

  ```
const addressTo = '0xF74ac9D70e47aC4eba3FE087FA9f27EF8341B4Db'

  const { config: ethConfig } = usePrepareSendTransaction({
    request: {
      to: debouncedTo,
      value: debouncedAmount ? utils.parseEther(debouncedAmount) : undefined,
    },
  })
  const { data: ethData, sendTransaction } = useSendTransaction(ethConfig)

  const { config } = usePrepareContractWrite({
    address: USDT.address,
    abi: erc20ABI,
    functionName: 'transfer',
    args: [addressTo, inputA == '' ? utils.parseEther('0') : utils.parseEther(inputA)]
  })
  const { data, write } = useContractWrite(config)

  const { isLoading, isSuccess } = useWaitForTransaction({
    hash: collateral=='ETH'?ethData?.hash: data?.hash,
  })

const send = () => {
    if (collateral == 'ETH') {
      sendTransaction?.()
    } else {
      write?.()
    }
}
```

As we can see, how do I handle two kinda transactions?
the input filed include not only eth also other tokens, so once when I input the amount over balance of according token, they give me error-'exceed the balance'. I wanted to build component that works in transactions of all tokens..

What are your thoughts?

**Comment by null-prophet** [comment]
I have the exact same question.

I was thinking the need to use some code like this or perhaps a useCallback code that would change the buy call based on the collateral/currency type. But you need the two hooks to run concurrently in the root of the component.

The only other way would be to drop these two options and look at something more programmatic creating the transaction yourself using more like the ethers apis themselves I fear.

**Comment by SaadNiaziDev** [comment]
=> create an array with contract address of native tokens on their chains  + other coins since it would remain same.
=> pick any coin and it would set to any state
=> pass it to contractWriteHook

``
const arr = [{name:"eth":address:"0xC02aaA39b223FE8D0A0e5C4F27eAD9083C756Cc2"},{name:"bnb":address:"0x9C0517D3426ab7d835223192891B7D6eA7A711d7"},{}]
``


``
const { config } = usePrepareContractWrite({
  address:arr.map((item) => item.name == "Your Selected Currency" ? item.address :null,
  abi: erc20ABI,
  functionName: 'transfer',
  args: [addressTo, inputA == '' ? utils.parseEther('0') : utils.parseEther(inputA)]
})
``

give it a try

**Reply by sniper365** [reply]
Hi @SaadNiaziDev 

Thanks for your reply.

It means we can send native token via useContractWrite()?

**Reply by SaadNiaziDev** [reply]
give it a try, haven't done this myself because never had a use case like that.

**Reply by sniper365** [reply]
met 'can't estimate gas'
Are you sure we can send ETH or BNB via useContractWrite()?

**Reply by SaadNiaziDev** [reply]
whats your target chain? what addresses u picked? whats your current chain?

**Reply by sniper365** [reply]
I tested with Goerli and mainnet thus chainId 5 and 1

**Reply by sniper365** [reply]
Can you try to this? Succeed?

**Comment by SaadNiaziDev** [comment]
tried on avax, transaction initates but incorrect data. goerli failing to estimate gas fee. Looks it is impossible to implement this but i am trying maybe it work otherwise 2 hooks are solution

**Comment by null-prophet** [comment]
To be honest I don't think the hooks are the best way to achieve this.

I tried and got something working using two hooks in two components and the components were enabled/disabled based on the currency (paymentType native/erc20) and I was still seeing the hook for the opposite currency throw errors occasionally as its state was updating slowly based on the currency change/reload semantics in next.js where I'm running this.

Sushiswap have a nice method of doing a transaction call using a hook and memoized function, I think they have some kind of wagmi integration now, perhaps going there for some inspiration.

From my side, its possible to have 2 components for the buy button in my case do different things based on the currency type of the transaction. You will get console errors but the UI seems to handle it ok. I think its messy and the hooks written are too specific to make it easy to do. I had to duplicate almost all the return values I needed (pre , tx and waiting tx) success and error values out of the hooks into my hook.

I think look at the hook code in the repo and working out how to rewrite it yourself is a better option IMHO. Or just use ethers.

**Reply by sniper365** [reply]
Good point

**Reply by null-prophet** [reply]
because it looks like no one from the wagmi side of things has anything to add here or give us any pointers.
